### ISSUE TITLE
MirroredStrategy incurring a lot of overheads as compared to standalone instance

### ISSUE THREAD

**Issue by shubham4060** [tensorflow/tensorflow#103564]
### Issue type

Performance

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.20

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04

### Python version

3.10

### CUDA/cuDNN version

12.2/9.14

### GPU model and memory

NVIDIA A100-SXM4-80GB (x8)

### Current behavior?

I have written a keras sequential model based on tensorflow. i had fine tuned it earlier for a single GPU but recently i was trying to run the same example with multi-GPU and i was observing a lot of overheads that i could not make sense of.

Model Details:

1. a sequential model with following layers 
     - dropout
     - dense with 16384 units (ReLU activation)
     - dense  with 8192 units (ReLU actibvation)
     - dropout
2. batch size of 4096, each row with around 11k features (tf.float32)
3. reading data from tfrecord files

For a single GPU i was able to achieve a step time of 60ms and verfied with tensorboard trace that prefetching was getting interleaved with computation. but when i ran the same model on multi-GPU (global batch size now of 4096*8), i observed many overheads and step time now increased to 170ms. My main concerns that i could spot from the tensorboard trace are

1. I used experimental_fetch_to_device to overlap fetch of batch data with computation, i could see some overlap in the trace but still for some portion of the MemcpyH2D, the gpu compute was blocked(and in some cases there was very minimal interleave). Doesn't the prefetch ensures that this memcpy is for the next step and therefore it should never have been blocked on memcpy in the first place? 
2. i also used distrbute_datasets_from_fn to create a distributed dataset and have used apply transformation to copy to the particular device(in addition to experimental_fetch_to_device as mentioned above), yet the memcpy is not getting properly interleaved. why is this happening?
3. I could a lot of gaps between the time each gpu started its compute, in some cases some gpu's were starting compute after the others have already finished which lead to a lot of time spent idle by other GPU's. why aren't all the GPUs starting compute at same time if trace shows MemcpyH2D for all finished at around same time?
4. even if i compare only compute time from the trace (minus the ncll all reduce gradient sync step), that also is much greater than the compute time in standalone instance. [tensorflow guide](https://www.tensorflow.org/guide/gpu_performance_analysis) this mentions that 

> When going from training with a single GPU to multiple GPUs on the same host, ideally you should experience the performance scaling with only the additional overhead of gradient communication and increased host thread utilization.

but as visible in the trace, even before ncll part, the compute took longer on any one GPU. 95ms in multiGPU vs 60ms in single GPU.
5. Pinned memory usage in tensorflow is not trivial, currently i referred to one of the existing tensorflow issues when mentioned to use batch after map pattern to use pinned memory, but i cant use .map().batch().cache() (i used for the above reported observations by keeping dataset small but asking for long term here) as it will then cache the memory in pinned memory which is what i ideally want but i am not able to do so due to something which seems like some hardcoded pinned memory limit (138GB) and so it gives OOM. i also tried increasing pagelocked mem limits in my OS but that still didn't help. any way on how i can increase the pinned memory so that i can make use of efficient Host to Device Transfers.
6. i also followed the the guide here but no luck. [Tensorflow guide](https://www.tensorflow.org/guide/gpu_performance_analysis)

have shared the sample code below which i used (might need to generate dummy tfrecord files first)
to run : `python3 train.py 4096 16384 - true true -`
also attached the multiGPU trace here :- [Trace](https://drive.google.com/drive/folders/1WFZRI8_e9S5tcNEbfpvLCaUK6CqH5CAX?usp=drive_link)

### Standalone code to reproduce the issue

```shell
# imports
import os
import sys
import argparse
import pickle
import logging
import json
from importlib import reload
import numpy as np
# if running on cpu
#os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
#os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ['TF_GPU_THREAD_MODE'] = 'gpu_private'
os.environ['TF_GPU_THREAD_COUNT'] = '128'
os.environ["TF_GPU_ALLOCATOR"] = "cuda_malloc_async"
import tensorflow as tf
print(tf.config.threading.get_inter_op_parallelism_threads())
print(tf.config.threading.get_intra_op_parallelism_threads())
from keras.utils import get_custom_objects
import tensorflow_probability as tfp
import datetime
import csv
from sklearn.model_selection import KFold

# cmdline args
BATCH_SIZE = int(sys.argv[1])
HIDDEN_LAYER = int(sys.argv[2])
TBOARD_DIR = sys.argv[3] if sys.argv[3] != '-' else None
USE_DATASET = (sys.argv[4].lower() == 'true')
USE_LARGE = (sys.argv[5].lower() == 'true')
PROFILE_BATCH_STR = sys.argv[6]
suf = sys.argv[7] if len(sys.argv) >= 8 else ''

from tensorflow.keras.callbacks import EarlyStopping
tf.config.experimental.enable_tensor_float_32_execution(True)
#tf.debugging.set_log_device_placement(True)
import logging

# 1️⃣ Create logger
logger = logging.getLogger("my_logger")
logger.setLevel(logging.DEBUG)  # Capture all levels

# 2️⃣ Create stream handler (stdout)
stream_handler = logging.StreamHandler()
stream_handler.setLevel(logging.DEBUG)

# 3️⃣ Create a formatter
formatter = logging.Formatter(
    "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
stream_handler.setFormatter(formatter)

# 4️⃣ Add handler to logger
logger.addHandler(stream_handler)

# tensorflow sanity

print("TensorFlow version:", tf.__version__)
print("GPUs available:", tf.config.list_physical_devices('GPU'))
gpus = tf.config.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

# model

loss_fn=None
activations = {
    'relu': tf.keras.layers.ReLU(),
    'leaky_relu': tf.keras.layers.LeakyReLU(alpha=0.05)
}

def keras_mse_loss(args):
    return tf.keras.losses.MeanSquaredError(
                reduction="sum_over_batch_size", name="mean_squared_error"
            )

def neural_network_l4(args):
    model = tf.keras.models.Sequential()
    if(args.get('in_dropout',0.1) != 0):
        model.add(tf.keras.layers.Dropout(rate=args.get('in_dropout',0.1)))
    for width in args.get('hidden_layers',[HIDDEN_LAYER,int(HIDDEN_LAYER/2)]):
        model.add(tf.keras.layers.Dense(units = width,
                                    use_bias=args.get('model_bias',False),
                                    activation=activations[args.get('activation','leaky_relu')],
                                    kernel_initializer=tf.keras.initializers.HeNormal(seed=args.get('seed',42)),
                                    kernel_regularizer=tf.keras.regularizers.L1L2(l1=args.get('l1_reg',1e-5), l2=args.get('l2_reg',1e-5))))
        if(args.get('batch_norm',0) == 1):
            model.add(tf.keras.layers.BatchNormalization())
        if(args.get('out_dropout',0.1) != 0):
            model.add(tf.keras.layers.Dropout(rate=args.get('out_dropout',0.1)))
    return model

def getModel(args):
    loss_fn = keras_mse_loss(args)
    optimizer = tf.keras.optimizers.Adam(learning_rate=args.get('lr',0.0001), clipvalue=args.get('clip_value',1))

    model = neural_network_l4({})
    model.compile(optimizer = optimizer,
                  loss = loss_fn,
                  metrics = [loss_fn])
    return model

args = {}

def parse_tfrecord(example_proto):
    feature_description = {
        "features": tf.io.FixedLenFeature([11000], tf.float32),
        "label": tf.io.FixedLenFeature([1], tf.float32),  # label is now float
    }
    parsed = tf.io.parse_single_example(example_proto, feature_description)
    return parsed["features"], parsed["label"]


strategy = tf.distribute.MirroredStrategy()
print("Number of devices:", strategy.num_replicas_in_sync)

def dataset_fn(input_context):
    with tf.device(f"/cpu:0"):
        device_id = input_context.input_pipeline_id
        dataset =  tf.data.TFRecordDataset(
                tf.io.gfile.glob(f"tfrecord_shards_parallel/dummy_data_{device_id}.tfrecord")).map(
                        parse_tfrecord, num_parallel_calls=tf.data.AUTOTUNE).batch(BATCH_SIZE).cache().apply(tf.data.experimental.copy_to_device(f'/gpu:{device_id}')).prefetch(4)
        options = tf.data.Options()
        options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.OFF
        return dataset.with_options(options)

def dataset_fn_test(input_context):
    with tf.device(f"/cpu:0"):
        device_id = input_context.input_pipeline_id
        dataset =  tf.data.TFRecordDataset(
                tf.io.gfile.glob(f"tfrecord_shards_parallel_test/dummy_data_0.tfrecord")).map(
                        parse_tfrecord, num_parallel_calls=tf.data.AUTOTUNE).batch(BATCH_SIZE).cache().apply(tf.data.experimental.copy_to_device(f'/gpu:{device_id}')).prefetch(tf.data.AUTOTUNE)
        options = tf.data.Options()
        options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
        return dataset.with_options(options)

input_options = tf.distribute.InputOptions(
    experimental_place_dataset_on_device=False,
    experimental_fetch_to_device=True,
    experimental_replication_mode=tf.distribute.InputReplicationMode.PER_REPLICA,
)

train_dist_data = strategy.distribute_datasets_from_function(
    dataset_fn, input_options
)

val_dist_data = strategy.distribute_datasets_from_function(
    dataset_fn_test, input_options
)

with strategy.scope():
    logger.info('configuring model...')
    model = getModel(args)
    logger.info('configured model')

    # model fit
    tb_args = {}
    if TBOARD_DIR:
        log_dir = f"{TBOARD_DIR}/logs_{BATCH_SIZE}_{HIDDEN_LAYER}{suf}/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
        tensorboard_cb = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1, profile_batch = PROFILE_BATCH_STR)
        tb_args["callbacks"] = [tensorboard_cb]
    model.fit(train_dist_data,
              epochs = 5,
              validation_data = val_dist_data, **tb_args)
```

### Relevant log output

```shell

```

**Comment by shubham4060** [comment]
@Venkat6871 @guptapriya can someone please help with this?

**Comment by Venkat6871** [comment]
Hi @shubham4060 ,
Apologies for the delay, and thanks for raising your concern here.
From your environment details, I noticed that you are running TensorFlow 2.20.0 with CUDA 12.2 / cuDNN 9.14.
This combination is not officially supported and may lead to performance or initialization issues.
According to the official [TensorFlow compatibility documentation](https://www.tensorflow.org/install/source#gpu) TensorFlow 2.20.0 is tested and built with:
```
CUDA 12.5
cuDNN 9.3
```
Could you please update your setup to use these versions and check if the issue persists?
If you still face the same behavior after aligning the environment, please share:
- A short TensorFlow Profiler trace (3–5 training steps)
- The outputs of nvidia-smi and nvcc --version

This information will help us investigate the performance overhead further.

Thank you!

**Comment by shubham4060** [comment]
Hi @Venkat6871 

Changed CUDA version to 12.5 and cuDNN to 9.3.

`nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Jun__6_02:18:23_PDT_2024
Cuda compilation tools, release 12.5, V12.5.82
Build cuda_12.5.r12.5/compiler.34385749_0`

nvidia-smi output

<img width="649" height="819" alt="Image" src="https://github.com/user-attachments/assets/4421909a-1833-44c7-a8c6-57f783b62700" />

Still getting the same observations, uploaded the trace here(can ignore step 25 of prior epoch here) [Trace](https://drive.google.com/drive/folders/1WFZRI8_e9S5tcNEbfpvLCaUK6CqH5CAX?usp=drive_link)
please check.

**Comment by Kayyuri** [comment]
Hi,
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.
Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
